filesystem/vdi.ksy

```diff
@@ -0,0 +1,228 @@
+        contents: [0x7F, 0x10, 0xDA, 0xBE]
+
+      - id: version
+        type: version
+
+      - id: header_size_optional
+        type: u4
+        if: subheader_size_is_dynamic
+        
+      - id: header_main
+        size: header_size
+        type: header_main
+    instances:
+      subheader_size_is_dynamic:
+        value: version.major>=1
+      header_size:
+        value: (subheader_size_is_dynamic?header_size_optional:336)
+      
+      bmap_size:
+        value: ((header_main.blocks_in_image * 4+header_main.geometry.sector_size-1) / header_main.geometry.sector_size)*header_main.geometry.sector_size
+      bmap_offset:
+        value: header_main.offset_bmap
+
+      blocks_offset:
+        -orig-id: data_offset
+        value: header_main.offset_data
+      block_size:
+        value: header_main.block_metadata_size + header_main.block_data_size
+
+    types:
+      uuid:
+        seq:
+          - id: uuid
+            size: 16
+      version:
+        seq:
+          - id: major
+            type: u2
+          - id: minor
+            type: u2
+      header_main:
+        seq:
+          - id: image_type
+            type: u4
+            enum: image_type
+
+          - id: image_flags
+            type: flags
+
+          - id: description
+            type: str
+            size: 256
+
+          - id: offset_bmap
+            type: u4
+            if: _parent.version.major>=1
+
+          - id: offset_data
+            type: u4
+            if: _parent.version.major>=1
+
+          - id: geometry
+            type: geometry
+
+          - id: reserved1
+            -orig-id: unused1
+            type: u4
+            if: _parent.version.major>=1
+
+          - id: disk_size
+            type: u8
+
+          - id: block_data_size
+            -orig-id: block_size
+            type: u4
+            doc: "Size of block (bytes)."
+
+          - id: block_metadata_size
+            -orig-id: block_extra
+            type: u4
+            if: _parent.version.major>=1
+
+          - id: blocks_in_image
+            type: u4
+
+          - id: blocks_allocated
+            type: u4
+
+          - id: uuid_image
+            type: uuid
+
+          - id: uuid_last_snap
+            type: uuid
+
+          - id: uuid_link
+            type: uuid
+
+          - id: uuid_parent
+            type: uuid
+            if: _parent.version.major>=1
+
+          - id: lchc_geometry
+            type: geometry
+            if: _parent.version.major>=1 and _io.pos + 16 <= _io.size
+          
+          #- id: reserved2
+          #  -orig_id: unused2
+          #  size: _io.size - _io.pos
+        types:
+          geometry:
+            seq:
+              - id: cylinders
+                type: u4
+
+              - id: heads
+                type: u4
+
+              - id: sectors
+                type: u4
+
+              - id: sector_size
+                type: u4
+          flags:
+            seq: #little endian assummed
```

Comment:
I believe VirtualBox is strictly little-endian anyway?

Reply:
The format is little endian, in qemu comments it is written that some fields are little endian even in memory.